### DISCUSSION TITLE
Jax with TFLite

### DISCUSSION THREAD

**Discussion by aizzaac** [jax-ml/jax#4504] [Q&A] [answered]
Hi

Can JAX be compressed with TFLite for use with embedded devices, mobile devices and microprocessors?

Thank you

**Comment by jekbradbury** [comment, marked as answer]
There's an experimental `jax2tf` converter that takes JAX functions and turns them into TF, and you could try running the result through the TFLite converter as well.

CC @gnecula

**Comment by gnecula** [comment]
As James said, TFLite is eventually within the scope that we have for jax2tf. At the moment, the conversion would likely fail because TFLite supports only a subset of the TF ops, and chances are that jax2tf will use ops outside that set. I'd be very interested to hear your experience if you try to use it, and perhaps we can work together to bridge the gap.

**Reply by aizzaac** [reply]
Sure. I will let you know. 
Also I have noticed that TRAX (which uses JAX as backend) can be exported to Keras [https://trax-ml.readthedocs.io/en/latest/notebooks/tf_numpy_and_keras.html#3.-Exporting-Trax-Models-for-Deployment](url) and eventually to TFLite.

**Reply by gnecula** [reply]
That conversion used TensorFlow.numpy, which I believe is less complete than jax2tf (not all primitives and no JAX transformation). The solution that the JAX team will support is based on jax2tf.

**Comment by cioc-shoreline** [comment]
As @jekbradbury recommended, we tried the workflow where we leverage jax2tf to convert from jax to tf and then use the TFLiteConverter to convert from tf to tflite. 

As @gnecula said, the conversion from tf to tflite did fail, specifically b/c of the TensorFlow XLA ops (e.g. lax.conv becomes tf.XlaConv). This caveat is documented here: https://github.com/google/jax/blob/master/jax/experimental/jax2tf/README.md#tensorflow-xla-ops

**Is there any guidance on how to proceed?** So far, I can think of the following:
1) Add custom operations to tflite converter and runtime. An issue is that the target must now have XLA, but this is fine for our environment. If this is the best path, would you be able to point me to an example of this?
1) Adjust the JAX code to avoid these operators. This feels hard as lax.conv is the basis for many functions and we also heavily use it explicitly. 
1) Somehow adjust the conversion within jax2tf to avoid these operators.
1) Do a by hand translation to tflite - possible, but fragile. 

Btw, Thank you all very much for JAX - fantastic software!

**Reply by gnecula** [reply]
I have an answer for #3. If you pass `--enable_xla=False` then the jax2tf convertor will try to avoid the XLA ops. The support for that mode is best-effort and at the moment only two such alterantive conversions are implemented, one for convolution and one for reduce_window. Give it a try, perhaps your convolution falls within the implemented case. If not, can you please construct a small repro case and then we can see if there is an easy way to convert that to the standard TF convolutions.

**Reply by cioc-shoreline** [reply]
Thank you! We will absolutely try out that param and see if we are able to make it work. I'll update you if it works or not. If not, I'll contain the sample repro code.

**Reply by 8bitmp3** [reply]
JAX to TensorFlow `jax2tf` is 🆒 🔥, thanks @gnecula. 

> Btw, Thank you all very much for JAX - fantastic software!

🙂

<img width="400" alt="image" src="https://user-images.githubusercontent.com/19637339/118399804-d2805300-b656-11eb-883f-7a7bb7469420.png">
